File: part-memo.clj
; Penyelesaian persoalan Partisi Bilangan bulat dengan Memoization

(def part
    (memoize
        (fn [[a b]]
            (cond 
              (or (< a 0) (< b 1)) 0
              (or (< a 2) (= b 1)) 1
              :else (+' (part [(- a b) b]) (part [a (dec b)]))))))
(defn integerPartition [n] (part [n n]))

(time (println (integerPartition 200)))

File: part-memo.py
import time
import sys

# Fungsi untuk memoize
def memoize(fn):
  memo = {}
  def iterMemo(a,b):
    if ((a,b) not in memo):
      memo[a,b] = fn(a,b)
    return memo[a,b]
  return iterMemo
  
@memoize
def part(a,b):
  if (a < 0 or b < 1):
    return 0
  elif (a < 2 or b == 1):
    return 1
  else:
    return (part(a-b,b) + part(a,b-1))
def integerPartition(n): return part(n,n)

sys.setrecursionlimit(3200)
start = time.time()
print(integerPartition(400))
end = time.time()
print(str((end - start)*1000) + " ms") # Waktu eksekusi
